### PR TITLE
NEP-18 Compatibility

### LINKED ISSUE #126: Support for numpy NaN functions: nan{min,max,mean,argmin,argmax,sum}

### LINKED ISSUE #396: Add support for median, average, and others
I've been trying to add support for more ufuncs that keep units, by adding items to `__copy_units`.  Unfortunately, many ufuncs start with calling `asarray` or `asanyarray`, and we probably don't want to mess with those: people calling `asarray` will not expect to get a `Quantity` in return.  Is there another way this could be achieved?

Relevant ufuncs include: `median`, `average`, `percentile`, `expand_dims`, `atleast_1d`, `atleast_2d`, `atleast_3d`, `tile`, `rot90`, and others.

### LINKED ISSUE #424: Calling the ufunc in __array_wrap__ is probably a bad idea
Since by that point, the ufunc has already run once. Although putting it in `__array_prepare__` may also not solve the problem, since that:

> `__array_prepare__` **should not attempt to access the array data** or resize the array, it is intended for setting the output array type, updating attributes and metadata, and performing any checks based on the input that may be desired before computation begins.

### LINKED ISSUE #547: Raise an error if a value returned by a wrapped function is a Quantity
Ps.- Only for values that are supposed to be wrapped (i.e. not those skipped).

### LINKED ISSUE #553: How to best deal with copyto and pint?
The `copyto` function complains that `argument must be numpy.ndarray, not Quantity`. I guess that this numpy function is one of the few that is not supported by pint currently. My current approach to work around this is:

```python    
# ...
tmp = destination.magnitude
np.copyto(tmp, source.magnitude)
destination= tmp * source.units
```

Are there any cleaner options to use `copyto` with units?

### LINKED ISSUE #617: Add numpy.interp to list of functions that don't work
numpy.interp doesn't work as expected for pint quantities.

### LINKED ISSUE #619: Pint's usage with Numpy's concatenate
```python
import numpy as np

from pint import UnitRegistry

ureg = UnitRegistry()

Q_ = ureg.Quantity

args = []

for k in range(0,17):
   uwnd = Q_(np.ones((1,73,144)),'meter / second')
    args.append(uwnd)
uwndTot = np.vstack(args)
print(uwndTot.shape)
for element in args:
   print(type(element)

```
I am using Pint and when  I  print the shape uwndTot I get
(17,73,144).

When I  print the values of element it turns out to be

<class 'pint.quantity.build_quantity_class.<locals>.Quantity'>

However if I iterate through uwndTot I get

```python
for element in uwndTot:
    print(type(element))
```
I get 

<class 'numpy.ndarray'>

I want to be able to use numpy's concatenate with Pint. Is that not possible ? I need the units of meters /second in the 3d dimensional array uwndTot. How do I do that ?

### LINKED ISSUE #682: __array_wrap__ broken for `insert`, etc.
The following code:
```python
import numpy as np
import pint
ureg = pint.UnitRegistry()
a = np.arange(3) * ureg.meter
np.insert(a, 2, 4 * ureg.meter)
```
produces:
```pytb
~/miniconda3/envs/py36/lib/python3.6/site-packages/numpy/lib/function_base.py in insert(arr, obj, values, axis)
   4439         new[tuple(slobj)] = arr[tuple(slobj2)]
   4440         if wrap:
-> 4441             return wrap(new)
   4442         return new
   4443     elif indices.size == 0 and not isinstance(obj, np.ndarray):

~/miniconda3/envs/py36/lib/python3.6/site-packages/pint/quantity.py in __array_wrap__(self, obj, context)
   1342 
   1343     def __array_wrap__(self, obj, context=None):
-> 1344         uf, objs, i_out = context
   1345 
   1346         # if this ufunc is not handled by Pint, pass it to the magnitude.

TypeError: 'NoneType' object is not iterable
```

I have no idea if `insert` has any hope of working right now, but that exception comes from code that is obviously wrong:

https://github.com/hgrecco/pint/blob/241bf7df1404e26d202df64e12e772d111803c31/pint/quantity.py#L1428-L1430

Here `context` defaults to `None` (as is [documented for numpy](https://docs.scipy.org/doc/numpy/reference/arrays.classes.html#numpy.class.__array_wrap__)), but the first line in `__array_wrap__` tries to unpack it unconditionally. I have no idea what the right behavior is in this case.

### LINKED ISSUE #700: UnitStrippedWarning when calling numpy ufuncs
Commit 4be294cf5621cf7278a05aab478de9936d787c9a introduced `UnitStrippedWarning`, and it is now extremely common for me to see this warning (I assume every time I use a ufunc).

For example, simply calling `np.cos` on a quantity will produce the warning. Is there some other way I'm supposed to use numpy ufuncs with quantities, or is this a bug?

If this is a limitation of the current ufunc implementation, is this something that the new [`__array_ufunc__`](https://docs.scipy.org/doc/numpy/reference/arrays.classes.html#numpy.class.__array_ufunc__) api can help with?

### LINKED ISSUE #790: The __array_function__ protocol for overriding NumPy's API
We have been drafting a new protocol to allow for overriding NumPy functions, e.g., for `np.concatenate`: http://www.numpy.org/neps/nep-0018-array-function-protocol.html

It would be great to experiment with getting this working in pint, since arrays with units are one of the intended use-cases.

### LINKED ISSUE #821: NumPy clip(), minimum(), maximum() broken
Running the code with `np.clip()`

```python
import pint
ureg = pint.UnitRegistry()
a = 3 * ureg.meter
b = 0 * ureg.meter
c = 2 * ureg.meter
import numpy as np
np.clip(a, b, c)
```

produces an error similar to #682.

```python-traceback
/usr/local/lib/python3.7/site-packages/pint/quantity.py:1377: UnitStrippedWarning: The unit of the quantity is stripped.
  warnings.warn("The unit of the quantity is stripped.", UnitStrippedWarning)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/usr/local/lib/python3.7/site-packages/numpy/core/fromnumeric.py in _wrapfunc(obj, method, *args, **kwds)
     55     try:
---> 56         return getattr(obj, method)(*args, **kwds)
     57

/usr/local/lib/python3.7/site-packages/pint/quantity.py in clip(self, first, second, out, **kwargs)
   1294
-> 1295         return self.__class__(self.magnitude.clip(**kwargs), self._units)
   1296

AttributeError: 'int' object has no attribute 'clip'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-7-1a75bcf59855> in <module>
----> 1 np.clip(a, b, c)

/usr/local/lib/python3.7/site-packages/numpy/core/fromnumeric.py in clip(a, a_min, a_max, out)
   1956
   1957     """
-> 1958     return _wrapfunc(a, 'clip', a_min, a_max, out=out)
   1959
   1960

/usr/local/lib/python3.7/site-packages/numpy/core/fromnumeric.py in _wrapfunc(obj, method, *args, **kwds)
     64     # a downstream library like 'pandas'.
     65     except (AttributeError, TypeError):
---> 66         return _wrapit(obj, method, *args, **kwds)
     67
     68

/usr/local/lib/python3.7/site-packages/numpy/core/fromnumeric.py in _wrapit(obj, method, *args, **kwds)
     48         if not isinstance(result, mu.ndarray):
     49             result = asarray(result)
---> 50         result = wrap(result)
     51     return result
     52

/usr/local/lib/python3.7/site-packages/pint/quantity.py in __array_wrap__(self, obj, context)
   1566
   1567     def __array_wrap__(self, obj, context=None):
-> 1568         uf, objs, i_out = context
   1569
   1570         # if this ufunc is not handled by Pint, pass it to the magnitude.

TypeError: cannot unpack non-iterable NoneType object
```

Might be unrelated, but `np.minimum()` and `np.maximum()` produces output without unit.

```pycon
>>> np.minimum(a, b)
0
>>> np.maximum(a, b)
3
```

### PR DESCRIPTION
**Pull request by jthielen**
Building off of the implementation of `__array_function__` in #764, this PR adds compatibility with NEP-18 in Pint (mostly in the sense of Quantity having `__array_function__` and being wrappable as a duck array; for Quantity wrapping other duck arrays, see #845). Many tests are added of NumPy functions being used with Pint Quantities by way of `__array_function__`.

Accompanying changes that were needed as a part of this implementation include:
- a complete refactor of `__array_ufunc__` and ufunc attribute fallbacks to work in parallel with `__array_function__`
- promoting `_eq` in `quantity` to `eq` in `compat`
- preliminary handling of array-like compatibility by defining upcast types and attempting to wrap and defer to all others (a follow-up PR, or set of PRs, will be needed to completely address #845 / #878)

Closes #126 
Closes #396 
Closes #424
Closes #547 
Closes #553
Closes #617
Closes #619
Closes #682 
Closes #700
Closes #764
Closes #790
Closes #821

### CODE REVIEW

**Comment by keewis** [comment]
thank you for working on this, @jthielen. Don't worry about the slow progress, as long as it is not completely on hold.

Running the xarray tests results in quite a few errors (probably more if we count the xfails). I can't do much right now, but I will try to investigate and report more this weekend.

**re** `inspect`: I already mentioned it in #764, but I strongly support dropping py2 (not my call, though).
Also, I'd say let's move adding tests for wrapping other NEP-18-compatible arrays to another PR so we can keep this as small as possible.

File: pint/quantity.py [hunk 1]
```diff
@@ -79,6 +78,262 @@ def wrapped(self, *args, **kwargs):
...
+
+# Handle single unit argument operations (axis ops, aggregations, etc.)
+for func_str in ['expand_dims', 'squeeze', 'rollaxis', 'moveaxis', 'fix', 'around', 'diagonal',
+                 'mean', 'ptp', 'ravel', 'round_', 'sort', 'median', 'nanmedian', 'transpose',
+                 'flip', 'copy', 'trim_zeros', 'average', 'nanmean']:
+    func = getattr(np, func_str)
+    @implements(func)
+    def implementation(a, *args, **kwargs):
+        (a,), output_wrap = unwrap_and_wrap_consistent_units(a)
+        return output_wrap(func(a, *args, **kwargs))
```
Comment (keewis) [inline pint/quantity.py]: there's something wrong with this and all similar closures below: at the end of quantity, all functions are implemented by wrapping `atleast_3d` (look at `HANDLED_FUNCTIONS[func].__globals__["func"]`).

I'm no expert on closures so I don't know why this happens or how to solve this (I will look into it, though). Does anyone else have any suggestions?

File: pint/quantity.py [hunk 2]
```diff
@@ -79,6 +78,262 @@ def wrapped(self, *args, **kwargs):
...
+for func_str in ['gradient', ]:
+    implement_func(func_str, input_units=None, output_unit='delta,div')
+
+for func_str in ['var', 'nanvar']:
+    implement_func(func_str, input_units=None, output_unit='variance')
+
+# TODO: broadcast_to (how to handle subok?)
+# TODO: 'amax', 'amin', 'nanmax', 'nanmin' (version dependent signatures)
+# TODO: nan_to_num (expected behavior with input values, sensible defaults?)
+
```
Comment (keewis) [inline pint/quantity.py]: Since it is being used a few times in xarray's codebase: can we also add a TODO for `result_type`?

File: pint/quantity.py (hunk 1 above)
Comment (keewis) [reply pint/quantity.py]: looks like creating `implementation` in a wrapper function (similar to `implement_func`) will solve this. We also should look at `HANDLED_FUNCTIONS[func].__closure__` and `HANDLED_FUNCTIONS[func].__code__.co_freevars` instead.

File: pint/quantity.py [hunk 3]
```diff
@@ -79,6 +78,262 @@ def wrapped(self, *args, **kwargs):
...
+    for arg in args_combo:
+        if isinstance(arg, Quantity):
+            out_units = arg.units
+        elif _is_quantity_sequence(arg):
+            out_units = arg[0].units
+        if out_units is not None:
+            break
+    return out_units
+    
+def convert_to_consistent_units(pre_calc_units=None, *args, **kwargs):
```
Comment (keewis) [inline pint/quantity.py]: it looks like `pre_calc_units` should be a keyword (-only?) argument, since it's easy to forget to pass `None` if we want to keep the default

File: pint/quantity.py [hunk 4]
```diff
@@ -79,6 +78,262 @@ def wrapped(self, *args, **kwargs):
...
+            fill_value.units)
+    else:
+        return (np.ones_like(a, dtype=dtype, order=order, subok=subok, shape=shape)
+                * fill_value)
+
+@implements(np.interp)
+def _interp(x, xp, fp, left=None, right=None, period=None):
+    # Need to handle x and y units separately
+    (x, xp, period), _ = unwrap_and_wrap_consistent_units(x, xp, period)
+    (fp, right, left), output_wrap = convert_to_consistent_units(fp, left, right)
```
Comment (keewis) [inline pint/quantity.py]: this looks like you swapped `unwrap_and_wrap_consistent_units` and `convert_to_consistent_units` (and maybe add `None` to the arguments of `convert_to_consistent_units`?)

**Comment by jthielen** [comment]
Thanks for the initial glace over the PR @keewis! In all my tinkering with this, I had forgotten to run the most recent set of changes past the test suite, which would have caught most of these. Whoops!

I've pushed some temporary fixups, which should get all the tests passing except the unimplemented ones. These changes may or may not be for naught after I redo the implementation logic using `inspect.signature`, which I hope to get to yet this weekend or early this coming week.

**Comment by keewis** [comment]
we should also keep in mind that with pint we need to make these changes optional: the tests without numpy installed currently fail because in that case `np` is `None`.

File: pint/testsuite/test_numpy.py [hunk 5]
```diff
@@ -154,34 +521,82 @@ def test_clip(self):
...
+    def test_nancumsum_numpy_func(self):
+        self.assertQuantityEqual(np.nancumsum(self.q_nan, axis=0), [[1, 2], [4, 2]] * self.ureg.m)
+
     def test_mean(self):
         self.assertEqual(self.q.mean(), 2.5 * self.ureg.m)
 
+    @helpers.requires_array_function_protocol()
+    def test_mean_numpy_func(self):
+        self.assertEqual(np.mean(self.q), 2.5 * self.ureg.m)
+        self.assertEqual(np.mean(q_temperature), self.Q_(2.5, self.ureg.m))
```
Comment (keewis) [inline pint/testsuite/test_numpy.py]: I guess that's the last error not related to missing implementations, except from the one related to `np.power`?
```suggestion
        self.assertEqual(np.mean(self.q_temperature), self.Q_(2.5, self.ureg.degC))
```

File: pint/compat/__init__.py [hunk 6]
```diff
@@ -66,6 +67,26 @@ def u(x):
...
+'To hide this warning use the following code to import pint:'
+"""
+import warnings
+with warnings.catch_warnings():
+    warnings.simplefilter("ignore")
+    import pint
+To disable the new behavior, see
+https://www.numpy.org/neps/nep-0018-array-function-protocol.html#implementation
+---
+""")
```
Comment (crusaderky) [inline pint/compat/__init__.py]: Wouldn't it be nicer to have the whole thing wrapped by ``"""``? (plus strip/rstrip/dedent as needed)

File: pint/compat/__init__.py [hunk 7]
```diff
@@ -66,6 +67,26 @@ def u(x):
...
 
+# TODO: remove this warning after v0.10
+class BehaviorChangeWarning(UserWarning):
+    pass
+_msg = ('The way pint handles numpy operations has changed with '
+'the implementation of NEP 18. Unimplemented numpy operations '
+'will now fail instead of making assumptions about units. Some '
+'functions, eg concat, will now return Quanties with units, '
+'where they returned ndarrays previously. See '
+'https://github.com/hgrecco/pint/pull/xxxx. '
```
Comment (crusaderky) [inline pint/compat/__init__.py]:
```suggestion
'https://github.com/hgrecco/pint/pull/905. '
```

File: pint/compat/__init__.py [hunk 8]
```diff
@@ -85,6 +106,23 @@ def _to_magnitude(value, force_ndarray=False):
...
+
+            np.concatenate([FakeArray()])
+            return True
+        except ValueError:
+            return False
+
+    HAS_NUMPY_ARRAY_FUNCTION = _test_array_function_protocol()
+
+    if HAS_NUMPY_ARRAY_FUNCTION:
+        warnings.warn(_msg, BehaviorChangeWarning)
```
Comment (crusaderky) [inline pint/compat/__init__.py]: This is going to annoy a lot of people who don't use pint with numpy but just happen to have numpy 1.17 installed in the same environment.
Could we issue the warning upon the first call to ``__array_function__``? Or upon the creation of the first Quantity with numpy data?

**Comment by hgrecco** [comment]
> **re** `inspect`: I already mentioned it in #764, but I strongly support dropping py2 (not my call, though).

See #906 but in principle I agree

File: pint/compat/__init__.py (hunk 8 above)
Comment (jthielen) [reply pint/compat/__init__.py]: Sounds good! I agree that would be annoying. Which of the two (warn on first `__array_function__` or warn on first creation) would make the most sense here? 

Also, would the following (for example in the warn-on-first-creation case) be a good replacement for the suggested way to hide the warning:

```python
import numpy as np
import warnings
from pint import Quantity

with warnings.catch_warnings():
    q = Quantity([1, 2, 3, 4], 'm')
```

or would something else be better?

**Comment by hgrecco** [comment]
As you sure have seen I have link a lot of issues to this PR. I am trying to make a list of numpy related issues that we should check again after this PR. Many of them will not be solved, but others might. The goal is to end 2019 with less open issues and PR!

**Comment by jthielen** [comment]
I've finally reached the point where all tests are passing locally with NumPy available, so I think that means @keewis we can move forward with the next phase of testing with xarray!

Also, I've updated the initial comment with the main areas I need to work on before I'd feel ready to remove WIP/draft status.

**Comment by keewis** [comment]
I opened a [new issue](https://github.com/pydata/xarray/issues/3594) where I listed the test failures. Most of these are probably issues with the code of xarray, but there are also a few missing function implementations: `nanpercentile`, `copyto`, `einsum`, `all`, `any`, `median`, `prod`, `sum`, `rank`, `isin`. Also `fabs`, `maximum` (and `minimum`?), but these are only used by the tests.

**Comment by keewis** [comment]
another method that currently falls back to the magnitude and causes lots of unexpected stripped units is `astype`.

**Comment by jthielen** [comment]
For the requested function implementations:

- Have been added:
  - `nanpercentile`
  - `copyto`
  - `isin`
  - `fabs`
  - `maximum`
  - `minimum`
- Should have already been there:
  - `median`
  - `sum`
- Need clarification:
  - `rank`: this is deprecated, so I don't think we should add it to new code?
- Can we leave as xfail for now to not hold up progress here?
  - `einsum`: determining the output unit for an arbitrary combination of sums and products from Einstein summation seems rather nontrivial
  - `all`, `any`: see #866 
  - `prod`: see https://github.com/hgrecco/pint/pull/764#issuecomment-523749008 / #867 

Also, the `astype` and `item` methods have been added.

**Comment by keewis** [comment]
Thanks.

> Should have already been there:

yes, that's my fault, I simply checked the test status for the aggregation tests. `sum` and `median` work properly with `int`, but not `float`.

> `rank`: this is deprecated, so I don't think we should add it to new code?

definitely. The warning never shows for me (or gets drowned in the flood of `UnitStrippedWarning`s), though, so I didn't realize.

**Comment by keewis** [comment]
somehow
```python
a = pint.Quantity([1, 3, 5], 'm')
a.ndim
```
raises a `AttributeError: NumPy method ndim was not callable.`, but `ndim` is a property (there is, however, a function named `numpy.ndim`)

File: pint/quantity.py [hunk 9]
```diff
@@ -1442,48 +1437,32 @@ def __ito_if_needed(self, to_units):
...
-                self._magnitude = _to_magnitude(self._magnitude, force_ndarray=True)
-                return getattr(self._magnitude, item)
-        elif item in self.__handled:
-            if not isinstance(self._magnitude, ndarray):
-                self._magnitude = _to_magnitude(self._magnitude, True)
-            attr = getattr(self._magnitude, item)
+                # TODO (#905 follow-up): Potentially problematic, investigate for duck arrays
+                magnitude_as_array = _to_magnitude(self._magnitude, force_ndarray=True)
+                return getattr(magnitude_as_array, item)
+        elif item in HANDLED_UFUNCS or self._wrapped_numpy_methods:
```
Comment (keewis) [inline pint/quantity.py]: I believe this should be
```python
        elif item in HANDLED_UFUNCS or item in self._wrapped_numpy_methods:
```
which would fix my problem with `ndim`

Comment (jthielen) [reply pint/quantity.py]: Yep, that is what I had intended. I had a commit incoming, but I thought I'd try bundling it with any other fixes after testing it with your test branch at https://github.com/pydata/xarray/pull/3600. However, I'm still a bit overwhelmed with trying to figure out where all the errors and `__array_interface__`/`__array_struct__` calls are coming from.

**Comment by jthielen** [comment]
Based on running through the xarray integration tests (as cleaned up in https://github.com/pydata/xarray/pull/3600), it looks like the following have issues that would be best resolved here:

- `maximum`
- `isin`
- `where`
- `sum` (in dataset aggregation test)

Also, there are many `UnitStrippedWarnings` raised in the xarray tests for the [`__array_interface__`](https://docs.scipy.org/doc/numpy/reference/arrays.interface.html#__array_interface__) and [`__array_struct__`](https://docs.scipy.org/doc/numpy/reference/arrays.interface.html#c.__array_struct__) attributes. What should the expected behavior from Pint be for these attributes/methods? The current behavior (call on magnitude and warn about stripped units), just call on magnitude without warning, or not implementing the attributes at all?

**Comment by keewis** [comment]
I looked at these issues and have a few comments / design questions:

the `maximum` issue is a problem with how xarray dispatches ufuncs (passing a data array as a second value works just fine), let's discuss that in pydata/xarray#3594.

how should `isin` work? Should it raise on incompatible units, or should it return `False`? I'd vote for not raising, since the different unit makes it a different value.

For `float` arrays, xarray's implementation of `where` uses `np.nan` to mask arrays. Same with xarray's nan-version of `sum`, which uses `where` to replace `nan` with `0`. Should `np.nan` be converted by pint's implementation of `where`, or should xarray be responsible to convert the default masking values? If xarray should change its code, what is the best way to convert without introducing special-cases for pint? `type(x)(np.nan)` creates a `dimensionless` quantity.

**Comment by jthielen** [comment]
@keewis Thanks for the input on those points! I've reimplemented `isin` based on your suggestion. Also, I think 0 and `np.nan` merit special behavior in Pint's implementation of `where`, for both xarray's purposes and general convenience for end users. I've re-implemented it accordingly, so hopefully that helps.

Also, `einsum` and a few other more functions have been added.

With this, I think all of xarray's needs are either met or intentionally deferred (e.g. `prod`/`any`/`all`)? If so, I think all that is left before removing draft status on this PR is further documenting and unit testing `numpy_func.py`, and resolving the [warning](https://github.com/hgrecco/pint/pull/905#discussion_r353427268) and [`__array_stuct__`/`__array_interface__`](https://github.com/hgrecco/pint/pull/905#issuecomment-562351799) discussions.

**Comment by keewis** [comment]
looks good to me. I didn't try `rot90` and `insert`, but `einsum` and `isin` work really well, with the exception of mixed args for `isin`: `np.isin(quantity, np.arange(5))` returns something different from the result of `np.isin(quantity, np.arange(5) * ureg.dimensionless)`. Is that intentional?

Also, `xarray` chose to use 0D arrays instead of scalars (so it passes `np.array(np.nan)` and `np.array(0)` to where). Should we consider these as scalars, too?

**Comment by keewis** [comment]
I just noticed we implement `__matmul__`, but not `Quantity.dot`, which will fall back to the ndarray method. Can we add that, too?

**Comment by jthielen** [comment]
> I didn't try `rot90` and `insert`, but `einsum` and `isin` work really well, with the exception of mixed args for `isin`: `np.isin(quantity, np.arange(5))` returns something different from the result of `np.isin(quantity, np.arange(5) * ureg.dimensionless)`. Is that intentional?

No, thanks for catching that, as I had forgotten the final branch of the conditional. It should be updated now.

> Also, `xarray` uses 0D arrays instead of scalars (so it passes `np.array(np.nan)` and `np.array(0)` to where). Should we consider these as scalars, too?

This shouldn't be a problem (I just changed the condition to testing for non-iterable non-quantity instead of scalar)

> I just noticed we implement `__matmul__`, but not `Quantity.dot`, which will fall back to the ndarray method. Can we add that, too?

Just added!

**Comment by hgrecco** [comment]
It would be good to summarize in the docs what has changed in relation to numpy support.

**Comment by jthielen** [comment]
Documentation and units tests have been added for `numpy_func.py`, and a summary of the changes in relation to NumPy support have been added to the docs. So, I think this is ready for a full review (bearing in mind the 2 unresolved conversations on warning placement and `__array_struct__`/`__array_interface__`), finally!

**Comment by hgrecco** [comment]
Awesome! I would like to merge this on wednesday. So I will kindly ask everyone who was involved to chime in by then.

**Comment by hgrecco** [comment]
By the way, on tueday it will be 7 years. 7 months and 7 days since I started working on Pint

**Comment by keewis** [comment]
I think we can leave the `__array_struct__` / `__array_interface__` issues for a new PR and until we find out what exactly causes them.

**Comment by hgrecco** [comment]
> I think we can leave the `__array_struct__` / `__array_interface__` issues for a new PR and until we find out what exactly causes them.

I agree.

File: pint/compat/__init__.py (hunk 8 above)
Comment (keewis) [reply pint/compat/__init__.py]: did we resolve this, yet? If not, I'd vote for raising on first creation since that would make sure people know that what they just created behaves different from before.

To silence it, I think the sample above works. However, I wouldn't assign a name to it (since that can make linters like `flake8` complain about a name being assigned but not used). We could also shorten it to `Quantity([], 'm')` or use `Quantity(np.array([]), "m")`?

Comment (jthielen) [reply pint/compat/__init__.py]: Since it seems like we have enough of a consensus, I've just pushed a commit to change the warning to on-first-creation.

**Comment by keewis** [comment]
did we resolve #882 yet?

**Comment by jthielen** [comment]
> did we resolve #882 yet?

No, not yet. I've implemented it here for now as in the first option ("Have `zeros_like`, `ones_like`, and `empty_like` return base zeros, ones, or empty ndarrays, respectively, and have `full_like` simply match the unit of the `fill_value` argument"), but that may be changed by how #882 is resolved (I'd personally still argue for this option, though, since it is more straightforward in interpretation and implementation).

**Comment by keewis** [comment]
okay, sounds good to me. I just wanted to make sure it's not forgotten.

@hgrecco, what's your opinion on this? Should we also close #882 with the merge, or do we need more discussion?

**Comment by hgrecco** [comment]
@keewis I do not have a strong opinion on this. I was inclined, as @jthielen said, the the units is part of the data (like dtype). So `*_like` should copy everything. But In any case, I would argue to  merge. I think we need to see this in action, close all resolved issues to see the pending ladscape and review the decision if needed.

**Comment by keewis** [comment]
I'm not sure if I understand you correctly: are you content with the current state or not?
In any case, +1 for merging and reviewing later.

**Comment by hgrecco** [comment]
In my opinon, the `*_like` should copy everything (including the units). But I have no other rationale than my gut feeling for now. In those cases, my motto is "let the person doing the hard work decide, and test if it feels right afterwards". So I will the implemented suggestion a try. I think what is right or wrong in these cases emerges from usage in real codebases.

File: docs/numpy.rst [hunk 10]
```diff
@@ -133,22 +137,31 @@ will be raised.
...
 usage. Quantities that must be converted to other units require additional
-memory and CPU cycles. On top of this, all `ufuncs` are implemented in the
-`Quantity` class by overriding `__array_wrap__`, a NumPy hook that is executed
-after the calculation and before returning the value. To our knowledge, there
-is no way to signal back to NumPy that our code will take care of the
-calculation. For this reason the calculation is actually done twice:
-first in the original ndarray and then in the one that has been
-converted to the right units. Therefore, for numerically intensive code, you
+memory and CPU cycles. Therefore, for numerically intensive code, you
 might want to convert the objects first and then use directly the magnitude.
```
Comment (hgrecco) [inline docs/numpy.rst]: Make a list to `wraps` here.

Comment (jthielen) [reply docs/numpy.rst]: A link has been added!

**Comment by hgrecco** [comment]
I done an initial review and it looks good so far. I still have to look at the tests in more detail. Please

**Comment by hgrecco** [comment]
I review the rest. Looks good to me. If the pending review is solved and no one raises an issue, I will merge it tomorrow.

**Comment by hgrecco** [comment]
@jthielen Sorry but I just realized I had not "submitted" the review. They are all minor things.

File: pint/quantity.py [hunk 11]
```diff
@@ -1302,62 +1315,56 @@ def __bool__(self):
...
-    __skip_other_args = 'ldexp multiply ' \
-                        'true_divide divide floor_divide fmod mod ' \
-                        'remainder'.split()
-
-    __handled = tuple(__same_units) + \
-                tuple(__require_units.keys()) + \
-                tuple(__prod_units.keys()) + \
-                tuple(__copy_units) + tuple(__skip_other_args)
+    # NumPy function/ufunc support
+    __array_priority__ = 17
```
[the remark this answers is not in the document]
Comment (jthielen) [reply pint/quantity.py]: While [the NumPy docs](https://docs.scipy.org/doc/numpy/reference/arrays.classes.html#numpy.class.__array_priority__) imply that it is no longer used for ufuncs when `__array_ufunc__` is in use, Dask, Sparse, and CuPy still all seem to declare it, so I thought it best to maintain it here. 

17 is simply the value Pint had previously. This puts it above ndarray (10), Dask (11), Sparse (12), and masked arrays (15), which is all good. However, CuPy is unfortunately at 100, so this may need to be increased if Pint Quantities should be above CuPy arrays. Would you want to increment this value now, or leave it at the current value of 17 and discuss any possible changes in #924?

File: pint/numpy_func.py [hunk 12]
```diff
@@ -0,0 +1,609 @@
...
+            and all(_is_quantity(item) for item in arg))
+
+
+def _get_first_input_units(args, kwargs={}):
+    """Obtain the first valid unit from a collection of args and kwargs."""
+    args_combo = list(args) + list(kwargs.values())
+    out_units=None
+    for arg in args_combo:
+        if _is_quantity(arg):
+            out_units = arg.units
```
[the remark this answers is not in the document]
Comment (jthielen) [reply pint/numpy_func.py]: Thank you for catching these! They were mostly carry-over from #764, which I should have reviewed more closely. I'll be pushing a commit soon with the updates.

File: pint/quantity.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (hgrecco) [reply pint/quantity.py]: Similar, to my previous comment. I think we should merge this ASAP to increase the base of tester. Regarding the number, I have two (not contradictory) options:
1. Kindly Suggest to CuPy to change the number
2. Get this value from an environmental variable (defaulting to 17). We should do this is in a way that is copyable by other packages (e.g. Naming the variable `PINT_NUMPY_ARRAY_PRIORITY` or similar)

**Comment by hgrecco** [comment]
I have just asked bors to merge this PR. I would like to thank each and everyone who has contributed to this PR, either by code, bugs, ideas, documentation and testing. This is a major step forward in Pint and a great example of collaborative work and open source development. 

Please review many Numpy related issues still open after this PR has been merged, closing those which are fixed and noting what is the new status after this PR.

**Comment by jthielen** [comment]
> Please review many Numpy related issues still open after this PR has been merged, closing those which are fixed and noting what is the new status after this PR.

I've gone through all the linked issues that remained open and either commented on their status or incorporated them into more general issues such as #845 and #925, so hopefully that helps!
